Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp

```diff
@@ -28,49 +28,381 @@
+                return std::nullopt;
+            continue;
+        }
+        
+        if (!isVisibleTextNode(*currentNode)) {
+            do {
+                if (auto newStart = nextInComposedTreeIgnoringUserAgentShadow(*currentNode)) {
+                    searchRange.start = { *newStart, 0 };
+                    currentNode = newStart;
+                } else
+                    return std::nullopt;
+            } while (is<DocumentType>(currentNode));
+            continue;
+        }
+        
+        auto& blockAncestor = nearestBlockAncestor(*currentNode);
+        Vector<RefPtr<Node>> textNodeList;
+        while (currentNode->isDescendantOf(blockAncestor) && is_lteq(treeOrder(BoundaryPoint(*currentNode, 0), searchRange.end))) {
+            if (currentNode->renderer() && is<Element>(currentNode) && currentNode->renderer()->style().isDisplayBlockLevel())
+                break;
+
+            if (isSearchInvisible(*currentNode)) {
+                if (auto newNode = nextInComposedTreeIgnoringUserAgentShadow(*currentNode))
+                    currentNode = newNode;
+                else
+                    break;
+                continue;
+            }
+            if (currentNode->isTextNode())
+                textNodeList.append(currentNode);
+            if (auto newNode = nextInComposedTreeIgnoringUserAgentShadow(*currentNode))
+                currentNode = newNode;
+            else
+                break;
+        }
+
+        if (auto range = findRangeFromNodeList(query, searchRange, textNodeList, wordStartBounded, wordEndBounded))
+            return range;
 
-    return resultRange.collapsed() ? std::nullopt : std::make_optional(resultRange);
+        if (!currentNode)
+            break;
+        
+        ASSERT_WITH_MESSAGE(is_gt(treeOrder(*currentNode, searchRange.start.container)), "Scroll To Text Fragment: Current node must follow start.");
+        
+        if (!is_gt(treeOrder(*currentNode, searchRange.start.container)))
+            return std::nullopt;
+        
+        searchRange.start = { *currentNode, 0 };
+    }
+    return std::nullopt;
+}
+
+// https://wicg.github.io/scroll-to-text-fragment/#next-non-whitespace-position
+static void advanceRangeStartToNextNonWhitespace(SimpleRange& range)
+{
+    while (!range.collapsed()) {
+        auto& node = range.startContainer();
+        auto offset = range.startOffset();
+        
+        // This check is not in the spec.
+        // I believe there is an error in the spec which I have filed an issue for
+        // https://github.com/WICG/scroll-to-text-fragment/issues/189
+        if (offset == node.length()) {
+            if (auto newStart = nextInComposedTreeIgnoringUserAgentShadow(node)) {
+                range.start = { *newStart, 0 };
+                continue;
+            }
+            return;
+        }
+        
+        if (isNonSearchableSubtree(node)) {
+            if (auto newStart = nextInComposedTreeIgnoringUserAgentShadow(node))
```

Comment:
I don't think this is right. composed tree isn't same thing as shadow-including tree order.
I suggest we skip shadow DOM stuff and just call regular NodeTraversal::nextSkippingChildren here.